### PR TITLE
Erase regions in writeback

### PR DESCRIPTION
**Pull request by matthewjasper**
Regions in `TypeckTables` (except canonicalized user annotations) are now erased. Further, we no longer do lexical region solving on item bodies with `-Zborrowck=mir`.

cc #68261
r? @nikomatsakis

### CODE REVIEW

**Comment by joelpalmer** [comment]
Triaged

File: src/librustc_infer/infer/mod.rs [hunk 1]
```diff
@@ -80,31 +80,45 @@ pub type Bound<T> = Option<T>;
 pub type UnitResult<'tcx> = RelateResult<'tcx, ()>; // "unify result"
 pub type FixupResult<'tcx, T> = Result<T, FixupError<'tcx>>; // "fixup result"
 
-/// A flag that is used to suppress region errors. This is normally
-/// false, but sometimes -- when we are doing region checks that the
-/// NLL borrow checker will also do -- it might be set to true.
-#[derive(Copy, Clone, Default, Debug)]
-pub struct SuppressRegionErrors {
-    suppressed: bool,
+#[derive(Copy, Clone, Debug)]
```
Comment (nikomatsakis) [inline src/librustc_infer/infer/mod.rs]: Can we add a doc comment explaining the overall role of this type?

Comment (nikomatsakis) [reply src/librustc_infer/infer/mod.rs]: I was imagining that we might give some coverage of how regions are handled in type-check vs NLL here. Though really we ought to explain this in a rustc-dev-guide section and link to *that*

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me with a comment, cc @matthewjasper

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/70066/checks?check_run_id=512947378, @bors r- (try rebasing).

File: src/test/ui/transmute/main.stderr [hunk 2]
```diff
@@ -4,8 +4,7 @@ error[E0512]: cannot transmute between types of different sizes, or dependently-
 LL |     transmute(x)
    |     ^^^^^^^^^
    |
-   = note: source type: `<C as TypeConstructor<'a>>::T` (size can vary because of <C as TypeConstructor>::T)
-   = note: target type: `<C as TypeConstructor<'b>>::T` (size can vary because of <C as TypeConstructor>::T)
+   = note: `<C as TypeConstructor>::T` does not have a fixed size
```
Comment (eddyb) [inline src/test/ui/transmute/main.stderr]: Wait, we don't allow transmuting between the same type with only lifetimes changed?
I think we do when it's beyond a pointer, even if we don't know if it's `Sized`.

**Comment by flip1995** [comment]
@matthewjasper Question regarding a Clippy rustup: rust-lang/rust-clippy#5333

Clippy has a check in the linting code that compares two types like this:

```
cx.tcx.expr_ty(expr1) == cx.tcx.expr_ty(expr2)
```

Before this PR, this returned `false` for `&'a T == &'b T`. Now it doesn't take the different lifetimes into account and returns `true`. Do you know of a way to compare two types taking their (user-defined) lifetimes into account?

**Comment by eddyb** [comment]
@flip1995 No, you were getting lucky before if that worked.
The only thing that can accurately reason about lifetimes is MIR borrowck.

File: src/test/ui/transmute/main.stderr (hunk 2 above)
Comment (matthewjasper) [reply src/test/ui/transmute/main.stderr]: We don't allow transmutes between type projections at all it seems:

```rust
pub trait TypeConstructor {
    type T;
}

unsafe fn transmute_lifetime<C>(x: C::T)
where C: TypeConstructor {
    transmute::<C::T, C::T>(x); // ERROR !
}
```

**Pull request merged**